src/main/java/use_cases/PlayerGameInteractor.java

```diff
@@ -0,0 +1,212 @@
+            if (currentGame.isValidCard(card)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+
+    /**
+     * Perform the logic required when a Player wins the Game, that is, set the Player as the Game winner,
+     * increment wins and losses for all Players
+     * @param player The player who has won the Game.
+     */
+    private void winLogic(Player player) {
+        currentGame.setWinner(player);
+        if(player instanceof HumanPlayer) {
+            ((HumanPlayer) player).incrementWins();
+        }
+        for(Player p: currentGame.getPlayers()) {
+            if(p instanceof HumanPlayer && !(p.equals(player))) {
+                ((HumanPlayer) p).incrementLosses();
+            }
+        }
+        currentGame.notifyGameObservers();
+    }
+
+    /**
+     * Check if the given Player is the winner of the Game
+     * @return True iff the Player has no Cards left in their Hand (played their last Card).
+     */
+    private boolean isWinner(Player player) {
+        return player.getCards().isEmpty();
+    }
+
+
+    /**
+     * This method implements all the logic for ComputerPlayers, including the logic for putting down a Card,
+     * picking one up, and skipping the turn.
+     * @param compPlayer The ComputerPlayer whose turn it is.
+     */
+    private void computerPlayerLogic(ComputerPlayer compPlayer) {
+        Card compCard = compPlayer.selectRandomCard(this.currentGame);
+        // If compCard is null, compPlayer has no valid cards, thus it must pick up and then try again.
+        if(compCard == null) {
+            compPlayer.pickUpCard(this.currentGame);
+            boolean hasAnyValid = anyValidCards(compPlayer);
+            if(hasAnyValid) {
+                // Player has a valid card, thus play it
+                Card toPlay = compPlayer.selectRandomCard(this.currentGame);
+                compPlayer.playCard(this.currentGame, toPlay);
+                if(isWinner(compPlayer)) {
+                    // The ComputerPlayer is the winner, set them as the Winner and notify the Game
+                    winLogic(compPlayer);
+                    return;
+                }
+                currentGame.changeCurrentTurn();
+                currentGame.notifyGameObservers();
+                return;
+            } else {
+                // Picked up but can't go, so skip turn
+                this.currentGame.notifyGameObservers();
+            }
+            currentGame.changeCurrentTurn();
+            currentGame.notifyGameObservers();
+        } else {
+            // Play the valid Card
+            compPlayer.playCard(this.currentGame, compCard);
+            if(isWinner(compPlayer)) {
+                // The ComputerPlayer is the winner, set them as the Winner and notify the Game
+                winLogic(compPlayer);
+                return;
+            }
+            currentGame.changeCurrentTurn();
+            currentGame.notifyGameObservers();
```

Comment:
Instead of my above comment, is it possible to call changeCurrentTurn and notifyGameObservers just once at the very end of the code? And instead, wherever you return from the if/else branches, "break" out of the branches, and go to the end of the code? That way, the code is less redundant.

Reply:
I could remove the nested else in the first if statement

So something like

If anyValid
	// play the card
	if isWinner
		do win logic
		return

	change turn and notify observers
Else
	same as before

Reply:
I was thinking along the lines of:

if anyvalid:
// play card
nested loop for if winner
// win logic (and return)
break; (for the entire outside if branch) [break statement guide](https://www.w3schools.com/java/java_break.asp)
else:
same as before
end if/else
change turn and notify observers

This way, you only need to have code to change turn and notify observers once, since in both branches either way you do both those things at the end.